### PR TITLE
feat(bundle): paginated bundle list

### PR DESCRIPTION
**Pull request by fredbi**
# Paginated bundle list

Signed-off-by: Frederic BIDON <frederic@oneconcern.com>

### Feature
This PR adds pagination support to the `bundle list` command. This means that we may query
large sets of keys while streaming back the output to the user.

### Requirements
Returned bundle metadata are sorted by key ID (ksuid).

### Description
I've built on top of PR #259 (parallel bundle metadata retrieval) and now decouple the retrieval of keys (which is sequential) from the retrieval of metadata (which is parallel).
Both queries progress asynchronously.
Each page (batch) of keys is submitted for parallel retrieval of bundle metadata, which are sorted before returning this batch.

### Main changes
The size of a page (batch size), expressed in # of keys, is an optional CLI  parameter.

The existing `pkg/core/ListBundle`  func keeps its interface (it is used by `pkg/web`).
I've added  the `pkg/core/ListBundleApply` func (now used by the CLI) to run any function asynchronously (that is in a non-blocking way) while fetching bundles.

Since this applied function may return an error, a signaling mechanism is put in place to abort the goroutines.

Some additional helper functions are now exported. The `ListBundleApply` func is constructed out of those.

I refactored error handling from storage (introducing sentinel errors in separate package) to avoid the string based error check in bundle list (en passant, made the adaptation for afero fs implem too).
### Testing
Besides CLI testing, the new feature is rather well covered by mocked tests, checking for no leaking goroutines, checking for most error paths.

### Suboptimal aspects
I am aware of the following suboptimal implementation issues:
* default values (batch size: 1024 or 20 kB, parallelism: 100) may not be the best ones. If users or benchmarkers find better defaults, I'll change them.
* the worker pool is spinned then wound down for every page. Although probably negligible for a few pages, this could add-up to some significant time with very large number of pages. However, this will be difficult to improve, given the sorting requirement.
* the error handling in the goroutines code is complex and required a lot of testing for blocking corner cases. This can be greatly improved by multiplexing values and errors on the same channel. At the moment, I am hesitating between  `chan struct{value: value_t, err: error}` and `chan func() (value_t, error)`. Your advice is welcome on this choice.

### CODE REVIEW

**Comment by ransomw1c** [comment]
> the retrieval of keys (which is sequential)

are we sure that what was intended by "pagination" wasn't in the sense of or according to the [page tokens](https://godoc.org/google.golang.org/api/iterator#PageInfo) such that that the list of pages of keys could be traversed sequentially while the keys themselves are being retrieved in parallel?

**Comment by fredbi** [comment]
@ransomw1c
Trying to clarify here: the pages of keys are retrieved sequentially. Bundle metadata attached to each key in a page (batch) are retrieved in parallel.

**Comment by fredbi** [comment]
See #267 and #268 for versions implementing channel multiplexing w/ errors.

File: Makefile [hunk 1]
```diff
@@ -192,7 +192,7 @@ mocks:
 
 .PHONY: runtests
 runtests: mocks
-	@go test ./...
+	@go test -v ./cmd/datamon//cmd
```
Comment (ransomw1c) [inline Makefile]: there are tests other than those in `cmd`

Comment (fredbi) [reply Makefile]: Yeah. Blunder. I’ve also seen it today. I unduly committed a local change just for me. Will revert

Comment (fredbi) [reply Makefile]: Yeah. Blunder. I’ve also seen it today. I unduly committed a local change just for me. Will revert

Comment (fredbi) [reply Makefile]: There is also an undue change in pkg/store errors: added an unused constant there.

Comment (fredbi) [reply Makefile]: There is also an undue change in pkg/store errors: added an unused constant there.

File: pkg/core/bundle_list.go [hunk 2]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+//   err := core.ListBundlesApply(repo, store, func(bundle model.BundleDescriptorOption) error {
+//				fmt.Fprintf(os.Stderr, "%v\n", bundle)
+//				return nil
+//			})
+func ListBundlesApply(repo string, store storage.Store, apply ApplyBundleFunc, opts ...BundleListOption) error {
+	var (
+		err, applyErr error
+		wg            sync.WaitGroup
+		workDone      bool
+		mutex         sync.Mutex
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: [the documentation](https://golang.org/pkg/sync/) mentions

> Other than the Once and WaitGroup types, most are intended for use by low-level library routines.

is there a way to communicate other than the shared `workDone` and lock?

Comment (fredbi) [reply pkg/core/bundle_list.go]: Mutex is commonplace.

I admit this comes a bit out of the nice full-channel picture... I thought about usIng a signalling channel again... in the end mutex expressed well my intent: first-past-the-post wins

File: pkg/core/bundle_list.go [hunk 3]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+			if !workDone {
+				doneChan <- struct{}{}
+			}
+			mutex.Unlock()
+			for range bundleChan {
+			} // wait for close
+			break
+		}
+	}
+	wg.Wait()
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: with the block on `close(bundleChan)`, the `wg sync.WaitGroup` is now superfluous

File: pkg/core/bundle_list.go [hunk 4]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+
+	// collect bundle metadata asynchronously
+	wg.Add(1)
+	go func(bundleChan chan<- model.BundleDescriptor, doneChan chan struct{}, wg *sync.WaitGroup) {
+		defer func() {
+			close(bundleChan)
+			wg.Done()
+		}()
+
+		bundlesChan, errChan, workers := ListBundlesChan(repo, store, append(opts, WithBundleDoneChan(doneChan))...)
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: `listBundlesChan` and `doSelectBundles`, perhaps?

File: pkg/core/bundle_list.go [hunk 5]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+
+		err = DoSelectBundles(bundlesChan, errChan, func(bundleBatch model.BundleDescriptors) {
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: supposing this is, in fact a package local function -- and perhaps even if it's not -- i'd favor discarding the callback passing and give write access to `bundleChan`

File: pkg/core/bundle_list.go [hunk 6]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+
+	workers.Wait()
+
+	return bundles, err // we may have some batches resolved before the error occurred
+}
+
+// ListBundlesChan returns a list of bundle descriptors from a repo. Each batch of returned descriptors
+// is sent on the output channel, following key lexicographic order.
+//
+// Simple use cases of this helper are wrapped in ListBundles (block until completion) and ListBundlesApply
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: let's remove the cruft:  if there's no dependency chain of `ListBundles` percolating to the ui, there's no need for `ListBundles`

Comment (ransomw1c) [reply pkg/core/bundle_list.go]: .. although it could be reasonable to add another commit to the mainline for the sake of having the bifurcated (sync/async) functionality in history

File: pkg/core/bundle_list.go [hunk 7]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+				bundleChan <- bundle // transfer a batch of metadata to the applied func
+			}
+		})
+
+		mutex.Lock()
+		workDone = true // nothing left to interrupt
+		close(doneChan)
+		mutex.Unlock()
+
+		workers.Wait()
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: again, prefer waiting on channels (with this `for range c {}` idiom) to a waitgroup when the two are equivalent.

reason:  less mutable state is simpler than less procedural data.

**Comment by ransomw1c** [comment]
note to self on "signalling channels" -- i.e. channels with type `struct{}` with a buffer size of one is that these channels convey two types doneness as encoded in `watchForInterrupts`
* "interrupts" wherein the channel is closed and empty
* non-"interrupts" wherein the channel has a value written to it

File: pkg/core/bundle_list.go [hunk 8]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+		close(outputChan)
+		wg.Done()
+	}()
+
+	var (
+		err                            error
+		errorChan1Open, errorChan2Open bool = true, true
+	)
+	for errorChan1Open || errorChan2Open { // wait on both channels to close
+		select {
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: [recall](https://golang.org/ref/spec#Select_statements)

> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection

so to prevent the enclosing `for` loop from being stochastic, i suggest adding a `switch` to read from precisely one or the other channel once one of the two is closed.

File: pkg/core/bundle_list.go [hunk 9]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+	)
+
+	for {
+		// get a batch of keys
+		ks, next, err = store.KeysPrefix(context.Background(), next, model.GetArchivePathPrefixToBundles(repo), "/", settings.bundleBatchSize)
+		if err != nil {
+			select {
+			case errorChan <- err:
+				break
+			case <-doneChan:
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: here, we don't preserve the meaning of "interrupt" in the sense of `watchForInterrupts` via, e.g.,

```
case _, interrupt <-doneChan:
	if interrupt { errorChan <- ErrInterrupted }
```

for some reason?

File: pkg/core/bundle_list.go [hunk 10]
```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
...
+		close(batchChan)
+		close(errorChan)
+		wg.Done()
+	}()
+
+	for {
+		select {
+		case <-doneChan:
+			errorChan <- ErrInterrupted
+			return
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: nitpick or possibly stylistic elucidation:  what `return` here where elsewhere `break`?

why this could be stylistic is that, for instance, i use double quotes for human readable strings in scripting languages like js and py and single quotes for keyword-like values.  is this a similar idiosyncratic convention where `return` consistently connotes a specific meaning wrt. programming patterns?

File: cmd/datamon/cmd/bundle_list.go [hunk 11]
```diff
@@ -7,14 +7,16 @@ import (
...
 
 	"github.com/spf13/cobra"
 )
 
+// BundleListCommand describes the CLI command for listing bundles
 var BundleListCommand = &cobra.Command{
 	Use:   "list",
 	Short: "List bundles",
-	Long:  "List the bundles in a repo",
+	Long:  "List the bundles in a repo, ordered by their key",
```
Comment (kerneltime) [inline cmd/datamon/cmd/bundle_list.go]: ```ordered by their bundle ID``` there is no entity in the model which is a key, that is an internal implementation.

File: cmd/datamon/cmd/bundle_list.go [hunk 12]
```diff
@@ -23,17 +25,18 @@ var BundleListCommand = &cobra.Command{
 		if err != nil {
 			logFatalln(err)
 		}
-		bundleDescriptors, err := core.ListBundles(params.repo.RepoName, remoteStores.meta, core.ConcurrentBundleList(params.core.ConcurrencyFactor))
-		if err != nil {
-			logFatalln(err)
-		}
-		for _, bd := range bundleDescriptors {
+		err = core.ListBundlesApply(params.repo.RepoName, remoteStores.meta, func(bundle model.BundleDescriptor) error {
```
Comment (kerneltime) [inline cmd/datamon/cmd/bundle_list.go]: I think this works in go but it is ugly to read.. can we define a function before we pass it?

File: cmd/datamon/cmd/cli_test.go [hunk 13]
```diff
@@ -493,12 +493,15 @@ func listBundles(t *testing.T, repoName string) (bundleListEntries, error) {
...
 		rle := bundleListEntry{
 			rawLine: line,
-			hash:    strings.TrimSpace(sl[0]),
+			hash:    strings.TrimSpace(sl[0]), // bundle key ID
 			message: strings.TrimSpace(sl[2]),
 			time:    t,
 		}
 		bles = append(bles, rle)
 	}
+	// bundles are ordered by lexicographic order of keys (ksuids).
```
Comment (kerneltime) [inline cmd/datamon/cmd/cli_test.go]: Drop ksuid here it is an internal implementation of the core package.
Also, use Bundle ID

File: cmd/datamon/cmd/flags.go [hunk 14]
```diff
@@ -143,11 +144,19 @@ func addCoreConcurrencyFactorFlag(cmd *cobra.Command) string {
...
+		"Heuristic on the amount of concurrency used by core operations (e.g. bundle list). "+
+			"Concurrent retrieval of bundle metadata is capped by the 'batch-size' parameter. "+
 			"Turn this value down to use less memory, increase for faster operations.")
 	return concurrencyFactor
 }
 
+func addBatchSizeFlag(cmd *cobra.Command) string {
+	batchSize := "batch-size"
+	cmd.Flags().IntVar(&params.core.BatchSize, batchSize, 1024,
+		"Number of bundle keys fetched per rountrip to storage -- keys are 20B long, defaults corresponds to 20kB pages")
```
Comment (kerneltime) [inline cmd/datamon/cmd/flags.go]: ```Number of bundles streamed to out together as a batch. This can be tuned for performance based on network connectivity```

File: pkg/core/bundle.go [hunk 15]
```diff
@@ -197,6 +203,31 @@ func ConcurrentBundleList(concurrentBundleList int) BundleListOption {
 	}
 }
 
+// BundleBatchSize sets the batch window to fetch bundle keys. It defaults to 100.
```
Comment (kerneltime) [inline pkg/core/bundle.go]: Drop ```It defaults to 100```.. ``` defaults to defaultBundleBatchSize```

File: cmd/datamon/cmd/bundle_list.go [hunk 16]
```diff
@@ -23,17 +25,18 @@ var BundleListCommand = &cobra.Command{
...
 			var buf bytes.Buffer
-			err := listLineTemplate.Execute(&buf, bd)
-			if err != nil {
-				log.Println("executing template:", err)
+			e := listLineTemplate.Execute(&buf, bundle)
+			if e != nil {
+				log.Println("executing template:", e)
```
Comment (kerneltime) [inline cmd/datamon/cmd/bundle_list.go]: This was not caught earlier but we should use fmt or change the logger to print to STDOUT. Right now this is printing to STDERR which makes it hard to script the output.

File: pkg/storage/status/status.go [hunk 17]
```diff
@@ -0,0 +1,12 @@
+// Package status declares error constants returned by the variou
```
Comment (kerneltime) [inline pkg/storage/status/status.go]: typo

File: pkg/storage/status/status.go [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+// implementations of the Store interface.
+package status
```
Comment (kerneltime) [inline pkg/storage/status/status.go]: Why not have it in the storage package?

File: pkg/core/bundle_list.go (hunk 2 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: Once will do it better.

File: pkg/core/bundle_list.go (hunk 4 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: Ok.

File: pkg/core/bundle_list.go (hunk 6 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: Used by web utility

File: pkg/storage/status/status.go (hunk 18 above)
Comment (fredbi) [reply pkg/storage/status/status.go]: @kerneltime because it creates a new constraint re not having cyclical dependencies. With a separate case package i am not adding any such implicit constraint on thid pkg

File: pkg/core/bundle_list.go (hunk 3 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: I disagree: although the bulk of the work is done, there may be some goroutines pending (waiting for doneChannels etc).

**However**, I agree that the Wait here is optional and we can explicitly allow for goroutine leak. This was not my initial choice (see test for leaked goroutines).

NOTE: usage is not strictly limited to CLI, I discovered we have a web server too. This one must not leak goroutines.

File: pkg/core/bundle_list.go (hunk 4 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: These funcs are typically what you would use to build the _porcelain_ layer you have been talking about. Agree that we can defer making them public to a more appropriate time.

File: pkg/core/bundle_list.go (hunk 5 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: ok to make DoSelect a private func.
You'll see in #267, #268 that it can be simplified to the point it could even disappear as a valuable helper.
Don't totally adhere to removing the callback. Need time to think about that. Thanks for raising this point.

File: pkg/core/bundle_list.go (hunk 7 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: ok with the principle but this is not equivalent

File: pkg/core/bundle_list.go (hunk 4 above)
Comment (ransomw1c) [reply pkg/core/bundle_list.go]: at this time, not pre-optimizing for lispification

File: cmd/datamon/cmd/bundle_list.go (hunk 11 above)
Comment (fredbi) [reply cmd/datamon/cmd/bundle_list.go]: ok

File: cmd/datamon/cmd/bundle_list.go (hunk 12 above)
Comment (fredbi) [reply cmd/datamon/cmd/bundle_list.go]: ok if that is your preference.

File: cmd/datamon/cmd/cli_test.go (hunk 13 above)
Comment (fredbi) [reply cmd/datamon/cmd/cli_test.go]: ok

File: cmd/datamon/cmd/flags.go (hunk 14 above)
Comment (fredbi) [reply cmd/datamon/cmd/flags.go]: ok

File: pkg/core/bundle.go (hunk 15 above)
Comment (fredbi) [reply pkg/core/bundle.go]: ok

File: pkg/core/bundle_list.go (hunk 10 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: break would just get us out from the `select`, not the loop as intended.

File: pkg/core/bundle_list.go [hunk 19]
```diff
@@ -57,37 +370,42 @@ func ListBundles(repo string, store storage.Store, opts ...BundleListOption) ([]
...
 
-	for i := 0; i < minInt(settings.concurrentBundleList, len(ks)); i++ {
+	// spin workers pool
+	for i := 0; i < minInt(settings.concurrentBundleList, len(keys)); i++ {
 		workers.Add(1)
 		go getBundleAsync(repo, store, keyChan, bundleChan, errorChan, &workers)
 	}
 
-	bds := make([]model.BundleDescriptor, 0, typicalBundlesNum) // preallocate some typical number of bundles, e.g. 1000
+	bds := make(model.BundleDescriptors, 0, settings.bundleBatchSize)
```
Comment (ransomw1c) [inline pkg/core/bundle_list.go]: noticed `len(keys)` instead of `bundleBatchSize` in #267 .. if this is to land as a separate commit, use `len(keys)` everywhere for consistency within `fetchBundleBatch`

File: cmd/datamon/cmd/bundle_list.go (hunk 16 above)
Comment (fredbi) [reply cmd/datamon/cmd/bundle_list.go]: sure. Will do.

**Comment by fredbi** [comment]
Closing this and taking the raised points in #267

File: pkg/storage/store.go [hunk 20]
```diff
@@ -25,6 +25,7 @@ const (
 	ErrNotSupported errString = "not supported"
 	ErrExists       errString = "exists already"
 	ErrObjectTooBig errString = "object too big to be read into memory"
+	ErrNotExists    errString = "object doesn't exist"
```
Comment (fredbi) [inline pkg/storage/store.go]: This one is superfluous.

File: pkg/core/bundle_list.go (hunk 3 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: ok understood. Removed unnessary waitgroup and wait here. Also removed explicit close of internal doneChannels, letting the gc do the cleaning work (documented this quirk as a comment).

File: pkg/core/bundle_list.go (hunk 5 above)
Comment (fredbi) [reply pkg/core/bundle_list.go]: taking advantage of this to move event types private in #267 (thus reducing the exposed symbols)
